File: src/main/java/higtools/modules/main/InvManager.java
package higtools.modules.main;

import higtools.HIGTools;
import meteordevelopment.meteorclient.events.world.TickEvent;
import meteordevelopment.meteorclient.gui.GuiTheme;
import meteordevelopment.meteorclient.gui.widgets.WWidget;
import meteordevelopment.meteorclient.gui.widgets.containers.WTable;
import meteordevelopment.meteorclient.gui.widgets.pressable.WButton;
import meteordevelopment.meteorclient.settings.IntSetting;
import meteordevelopment.meteorclient.settings.Setting;
import meteordevelopment.meteorclient.settings.SettingGroup;
import meteordevelopment.meteorclient.settings.StringSetting;
import meteordevelopment.meteorclient.systems.modules.Module;
import meteordevelopment.meteorclient.utils.render.MeteorToast;
import meteordevelopment.orbit.EventHandler;
import net.minecraft.item.Items;
import net.minecraft.registry.Registries;
import net.minecraft.screen.slot.SlotActionType;
import net.minecraft.util.Identifier;

import java.util.Arrays;
import java.util.stream.IntStream;

import static higtools.utils.HIGUtils.csToPs;


public class InvManager extends Module {
    private final Identifier[] itemIds = IntStream.range(0, 9)
        .mapToObj(i -> new Identifier(""))
        .toArray(Identifier[]::new);

    private final SettingGroup sgGeneral = settings.getDefaultGroup();
    private final SettingGroup slotsGroup = settings.createGroup("Slots", false);

    private final Setting<String> slot0 = slotsGroup.add(new StringSetting.Builder()
        .name("slot0")
        .defaultValue("")
        .visible(() -> false)
        .onChanged(cur -> {
            itemIds[0] = new Identifier(cur);
            System.out.printf("<HIGTools/InvManager> Loaded %s into slot0\n", itemIds[0]);
        })
        .build()
    );

    private final Setting<String> slot1 = slotsGroup.add(new StringSetting.Builder()
        .name("slot1")
        .defaultValue("")
        .visible(() -> false)
        .onChanged(cur -> {
            itemIds[1] = new Identifier(cur);
            System.out.printf("<HIGTools/InvManager> Loaded %s into slot1\n", itemIds[1]);
        })
        .build()
    );

    private final Setting<String> slot2 = slotsGroup.add(new StringSetting.Builder()
        .name("slot2")
        .defaultValue("")
        .visible(() -> false)
        .onChanged(cur -> {
            itemIds[2] = new Identifier(cur);
            System.out.printf("<HIGTools/InvManager> Loaded %s into slot2\n", itemIds[2]);
        })
        .build()
    );
    private final Setting<String> slot3 = slotsGroup.add(new StringSetting.Builder()
        .name("slot3")
        .defaultValue("")
        .visible(() -> false)
        .onChanged(cur -> {
            itemIds[3] = new Identifier(cur);
            System.out.printf("<HIGTools/InvManager> Loaded %s into slot3\n", itemIds[3]);
        })
        .build()
    );
    private final Setting<String> slot4 = slotsGroup.add(new StringSetting.Builder()
        .name("slot4")
        .defaultValue("")
        .visible(() -> false)
        .onChanged(cur -> {
            itemIds[4] = new Identifier(cur);
            System.out.printf("<HIGTools/InvManager> Loaded %s into slot4\n", itemIds[4]);
        })
        .build()
    );
    private final Setting<String> slot5 = slotsGroup.add(new StringSetting.Builder()
        .name("slot5")
        .defaultValue("")
        .visible(() -> false)
        .onChanged(cur -> {
            itemIds[5] = new Identifier(cur);
            System.out.printf("<HIGTools/InvManager> Loaded %s into slot5\n", itemIds[5]);
        })
        .build()
    );
    private final Setting<String> slot6 = slotsGroup.add(new StringSetting.Builder()
        .name("slot6")
        .defaultValue("")
        .visible(() -> false)
        .onChanged(cur -> {
            itemIds[6] = new Identifier(cur);
            System.out.printf("<HIGTools/InvManager> Loaded %s into slot6\n", itemIds[6]);
        })
        .build()
    );
    private final Setting<String> slot7 = slotsGroup.add(new StringSetting.Builder()
        .name("slot7")
        .defaultValue("")
        .visible(() -> false)
        .onChanged(cur -> {
            itemIds[7] = new Identifier(cur);
            System.out.printf("<HIGTools/InvManager> Loaded %s into slot7\n", itemIds[7]);
        })
        .build()
    );
    private final Setting<String> slot8 = slotsGroup.add(new StringSetting.Builder()
        .name("slot8")
        .defaultValue("")
        .visible(() -> false)
        .onChanged(cur -> {
            itemIds[8] = new Identifier(cur);
            System.out.printf("<HIGTools/InvManager> Loaded %s into slot8\n", itemIds[8]);
        })
        .build()
    );

    private final Setting<Integer> delay = sgGeneral.add(new IntSetting.Builder()
        .name("delay")
        .description("Delay in ticks between moving items.")
        .defaultValue(1)
        .range(1, 35)
        .sliderRange(1, 35)
        .build()
    );

    public InvManager() {
        super(HIGTools.MAIN, "InvManager", "Automatically organize your inventory.");
    }

    @Override
    public void onActivate() {
        super.onActivate();
        System.out.printf("<HIGTools/InvManager> %s%n", Arrays.toString(itemIds));
    }

    @Override
    public WWidget getWidget(GuiTheme theme) {
        WTable table = theme.table();
        fillTable(theme, table);

        return table;
    }

    private void fillTable(GuiTheme theme, WTable table) {
        WButton save = table.add(theme.button("Save")).expandCellX().right().widget();
        save.action = this::save;

        WButton reset = table.add(theme.button("Reset")).right().widget();
        reset.action = this::reset;
    }

    @EventHandler
    public void tick(TickEvent.Pre event) {
        if (mc.player.age % delay.get() != 0) return;

        for (int i = 0; i <= 8; i++) {
            if (itemIds[i].toString().replace("minecraft:", "").equals("")) continue;
            if (!Registries.ITEM.getId(mc.player.getInventory().getStack(i).getItem()).equals(itemIds[i])) {
                for (int j = 9; j <= 35; j++) {
                    if (Registries.ITEM.getId(mc.player.getInventory().getStack(j).getItem()).equals(itemIds[i])) {
                        mc.interactionManager.clickSlot(mc.player.currentScreenHandler.syncId, csToPs(j), 0, SlotActionType.PICKUP, mc.player);
                        mc.interactionManager.clickSlot(mc.player.currentScreenHandler.syncId, csToPs(i), 0, SlotActionType.PICKUP, mc.player);
                        mc.interactionManager.clickSlot(mc.player.currentScreenHandler.syncId, csToPs(j), 0, SlotActionType.PICKUP, mc.player);
                        mc.interactionManager.tick();
                        break;
                    }
                }
            }
        }
    }

    private void reset() {
        slot0.set("");
        slot1.set("");
        slot2.set("");
        slot3.set("");
        slot4.set("");
        slot5.set("");
        slot6.set("");
        slot7.set("");
        slot8.set("");

        mc.getToastManager().add(new MeteorToast(Items.ENDER_CHEST, "InvManager", "Cleared Saved Hotbar"));
    }

    private void save() {
        if (mc.player == null) return;

        slot0.set(Registries.ITEM.getId(mc.player.getInventory().getStack(0).getItem()).toString());
        slot1.set(Registries.ITEM.getId(mc.player.getInventory().getStack(1).getItem()).toString());
        slot2.set(Registries.ITEM.getId(mc.player.getInventory().getStack(2).getItem()).toString());
        slot3.set(Registries.ITEM.getId(mc.player.getInventory().getStack(3).getItem()).toString());
        slot4.set(Registries.ITEM.getId(mc.player.getInventory().getStack(4).getItem()).toString());
        slot5.set(Registries.ITEM.getId(mc.player.getInventory().getStack(5).getItem()).toString());
        slot6.set(Registries.ITEM.getId(mc.player.getInventory().getStack(6).getItem()).toString());
        slot7.set(Registries.ITEM.getId(mc.player.getInventory().getStack(7).getItem()).toString());
        slot8.set(Registries.ITEM.getId(mc.player.getInventory().getStack(8).getItem()).toString());

        mc.getToastManager().add(new MeteorToast(Items.ENDER_CHEST, "InvManager", "Saved Hotbar"));
    }
}

File: src/main/kotlin/higtools/modules/player/InvManager.kt
package higtools.modules.player

import higtools.*
import higtools.modules.HIGTools
import meteordevelopment.meteorclient.events.world.TickEvent
import meteordevelopment.orbit.EventHandler
import net.minecraft.screen.slot.SlotActionType
import net.minecraft.text.Text
import net.minecraft.util.Formatting
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry

class InvManager:MeteorModule(HIGTools.HIG, "InvManager", "Automatically organize your inventory") {

    private val slotsGroup = settings.createGroup("Slots", false)

    private var slot0 by slotsGroup.add(SValue("Slot0",
                                               "",
                                               "",
                                               { false },
                                               { cur -> itemIds[0] = Identifier(cur).also { println("<Llama/InvManager> Loaded $it into slot0") } }))
    private var slot1 by slotsGroup.add(SValue("Slot1",
                                               "",
                                               "",
                                               { false },
                                               { cur -> itemIds[1] = Identifier(cur).also { println("<Llama/InvManager> Loaded $it into slot1") } }))
    private var slot2 by slotsGroup.add(SValue("Slot2",
                                               "",
                                               "",
                                               { false },
                                               { cur -> itemIds[2] = Identifier(cur).also { println("<Llama/InvManager> Loaded $it into slot2") } }))
    private var slot3 by slotsGroup.add(SValue("Slot3",
                                               "",
                                               "",
                                               { false },
                                               { cur -> itemIds[3] = Identifier(cur).also { println("<Llama/InvManager> Loaded $it into slot3") } }))
    private var slot4 by slotsGroup.add(SValue("Slot4",
                                               "",
                                               "",
                                               { false },
                                               { cur -> itemIds[4] = Identifier(cur).also { println("<Llama/InvManager> Loaded $it into slot4") } }))
    private var slot5 by slotsGroup.add(SValue("Slot5",
                                               "",
                                               "",
                                               { false },
                                               { cur -> itemIds[5] = Identifier(cur).also { println("<Llama/InvManager> Loaded $it into slot5") } }))
    private var slot6 by slotsGroup.add(SValue("Slot6",
                                               "",
                                               "",
                                               { false },
                                               { cur -> itemIds[6] = Identifier(cur).also { println("<Llama/InvManager> Loaded $it into slot6") } }))
    private var slot7 by slotsGroup.add(SValue("Slot7",
                                               "",
                                               "",
                                               { false },
                                               { cur -> itemIds[7] = Identifier(cur).also { println("<Llama/InvManager> Loaded $it into slot7") } }))
    private var slot8 by slotsGroup.add(SValue("Slot8",
                                               "",
                                               "",
                                               { false },
                                               { cur -> itemIds[8] = Identifier(cur).also { println("<Llama/InvManager> Loaded $it into slot8") } }))

    private var delay by mainGroup.add(IValue("Delay", 15, "Delay to use", 1..35, 1))
    private var reset by mainGroup.add(BValue("Reset", false, "Toggle this to reset, too lazy to make proper thing", { true }, { reset() }))

    private var itemIds:Array<Identifier> = Array(9) { Identifier("") }

    override fun onActivate() {
        super.onActivate()
        println("<HIGTools/InvManager> ${itemIds.contentToString()}")
    }

    @EventHandler
    fun tick(event:TickEvent.Pre) {
        if (mc.player!!.age % delay == 0) {
            for (i in 0..8) {
                if (itemIds[i].toString().replace("minecraft:", "") == "") continue
                if (Registry.ITEM.getId(mc.player!!.inventory.getStack(i).item) != itemIds[i]) {
                    for (j in 9..35) {
                        if (Registry.ITEM.getId(mc.player!!.inventory.getStack(j).item) == itemIds[i]) {
                            mc.interactionManager!!.clickSlot(mc.player!!.currentScreenHandler.syncId,
                                                              csToPs(j),
                                                              0,
                                                              SlotActionType.PICKUP,
                                                              mc.player)
                            mc.interactionManager!!.clickSlot(mc.player!!.currentScreenHandler.syncId,
                                                              csToPs(i),
                                                              0,
                                                              SlotActionType.PICKUP,
                                                              mc.player)
                            mc.interactionManager!!.clickSlot(mc.player!!.currentScreenHandler.syncId,
                                                              csToPs(j),
                                                              0,
                                                              SlotActionType.PICKUP,
                                                              mc.player)
                            mc.interactionManager!!.tick()
                            break
                        }
                    }
                }
            }
        }
    }

    private fun reset() {
        slot0 = ""
        slot1 = ""
        slot2 = ""
        slot3 = ""
        slot4 = ""
        slot5 = ""
        slot6 = ""
        slot7 = ""
        slot8 = ""
    }

    private fun save() {
        slot0 = Registry.ITEM.getId(mc.player!!.inventory.getStack(0).item).toString()
        slot1 = Registry.ITEM.getId(mc.player!!.inventory.getStack(1).item).toString()
        slot2 = Registry.ITEM.getId(mc.player!!.inventory.getStack(2).item).toString()
        slot3 = Registry.ITEM.getId(mc.player!!.inventory.getStack(3).item).toString()
        slot4 = Registry.ITEM.getId(mc.player!!.inventory.getStack(4).item).toString()
        slot5 = Registry.ITEM.getId(mc.player!!.inventory.getStack(5).item).toString()
        slot6 = Registry.ITEM.getId(mc.player!!.inventory.getStack(6).item).toString()
        slot7 = Registry.ITEM.getId(mc.player!!.inventory.getStack(7).item).toString()
        slot8 = Registry.ITEM.getId(mc.player!!.inventory.getStack(8).item).toString()
        mc.inGameHud.chatHud.addMessage(Text.of("<${Formatting.LIGHT_PURPLE}Llama/InvManager${Formatting.RESET}> Saved inventory -> ${
            itemIds.contentToString().replace("minecraft:", "")
        }"))
    }

}
